<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>largeQvalue by abrown25</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>largeQvalue</h1>
        <h2>A program for calculating FDR estimates with large datasets</h2>
        <a href="https://github.com/abrown25/qvalue" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h2>
<a id="largeqvalue" class="anchor" href="#largeqvalue" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>largeQvalue:</h2>

<p>This is an implementation of the qvalue package (Alan Dabney, John D. Storey and with assistance from Gregory R. Warnes (). qvalue: Q-value estimation for false discovery rate control. R package version 1.34.0.) which is designed for use with large datasets where memory or computation time may be an issue with R. It has been used to analyse full cis scans of gene expression data, with hundreds of millions of P values. A description of the algorithms and instructions for usage can be found in the accompanying paper: <a href="http://biorxiv.org/content/early/2014/10/06/010074">http://biorxiv.org/content/early/2014/10/06/010074</a>. </p>

<p>Given a whitespace separated table (specified by --input or as the last option on the command line) and a column number (--col, default = 1), the program will write a new file where the equivalent q values are in the last column. The --param flag specifies a file to write parameter values used to estimate these q values, this file can be copied and pasted into R to produce diagnostic plots.</p>

<h3>
<a id="binaries" class="anchor" href="#binaries" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Binaries</h3>

<p>The latest binary can be found in the bin folder on this site. An older version can be found here:</p>

<p>ftp://ftp.sanger.ac.uk/pub/resources/software/largeqvalue/largeQvalue.tar.gz</p>

<h3>
<a id="example-commands" class="anchor" href="#example-commands" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Example commands:</h3>

<pre><code>largeQvalue &lt; QTLresults.txt
</code></pre>

<p>The file QTLresults.txt contains the p values for association in the first column, the results are written to stdout with q values controlling for multiple testing, calculated using default settings, in a new final column.</p>

<pre><code>largeQvalue --header --col 4 --out output --param parameter_file --lambda 0,0.9,0.05 --robust QTLresults.txt
</code></pre>

<p>The p values can be found in the 4th column of QTLresults.txt, write the results to output and the estimated parameter values to parameter_file. Use values of lambda from 0 to 0.9 in steps of 0.05 to estimate proportion of null hypotheses (standard settings in qvalue) and produce estimates of q values robust for small p values.</p>

<h3>
<a id="list-of-options" class="anchor" href="#list-of-options" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>List of options:</h3>

<pre><code>    Usage: largeQvalue [options]

    Options:

    --help    Print help and quit.

    --version Print version and quit.

    --input CHAR
            File  containing  p  values  to analyse. This can also be specified by the
            last argument on the command-line after all others have been parsed. If
            neither are present, it is taken from the stdin [stdin].

    --out CHAR
             File to write results to [stdout].

    --param CHAR
             Print out parameter list to specified file.

    --header  Input has header line [FALSE].

    --col INT Column with p values [1].

    --sep CHAR
             Separator to use to separate the column with q values. Specified as either
             space or tab (which can be shortened to s or t) [tab].

    --issorted
             File has already been sorted with no missing values [FALSE].

    --pi0 DOUBLE
             Use given value of pi0.

    --lambda DOUBLE(,DOUBLE,DOUBLE)
             Either a fixed number or a sequence given as 0,0.9,0.05 (start,end,step)
             [0,0.9,0.05].

    --robust  More robust values for small p values [FALSE].

    --df DOUBLE
              Number of degrees of freedom used by the spline when estimating pi0 [3].

    --log     Smoothing spline applied to log pi0 values [FALSE].

    --boot    Apply bootstrap method to find pi0 [FALSE].

    --seed DOUBLE
              Set seed for generating bootstrap samples [0].

    --fast DOUBLE
              Report nominal P value threshold for each gene corresponding to given
              FDR threshold when input is a fastQTL results file.

</code></pre>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/abrown25/qvalue/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/abrown25/qvalue/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/abrown25/qvalue"></a> is maintained by <a href="https://github.com/abrown25">abrown25</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
